### PR TITLE
feat: allow exporting memory file and trace file in non proof mode

### LINKED ISSUE #472: Export trace and memory in execution mode
Allow exporting the trace and memory when running in execution mode (non-proof mode) on the CLI. It is currently only doable with proof mode

### PR DESCRIPTION
**Pull request by TAdev0**
Resolves #472

### CODE REVIEW

File: pkg/vm/vm.go [hunk 1]
```diff
@@ -537,7 +523,7 @@ func (vm *VirtualMachine) updateFp(instruction *a.Instruction, dstAddr *mem.Memo
 	}
 }
 
-func (vm *VirtualMachine) relocateTrace() []Trace {
+func (vm *VirtualMachine) RelocateTrace() []Trace {
```
Comment (rodrigo-pino) [inline pkg/vm/vm.go:526]: Since we are making it public I'd suggest adding a description to it

File: pkg/vm/vm.go [hunk 2]
```diff
@@ -136,9 +133,7 @@ func (vm *VirtualMachine) RunStep(hintRunner HintRunner) error {
 	}
 
 	// store the trace before state change
-	if vm.config.ProofMode {
-		vm.Trace = append(vm.Trace, vm.Context)
-	}
+	vm.Trace = append(vm.Trace, vm.Context)
```
Comment (rodrigo-pino) [inline pkg/vm/vm.go:136]: I wouldn't add the trace every time. Collecting the trace vs not is a performance hit. I would suggest to have a flag in the VM.config (perhaps `get_trace`, `collect_trace` or any name you think is best) and keep the if

File: pkg/vm/vm.go [hunk 3]
```diff
@@ -91,11 +91,8 @@ type VirtualMachine struct {
 func NewVirtualMachine(
 	initialContext Context, memory *mem.Memory, config VirtualMachineConfig,
 ) (*VirtualMachine, error) {
-	// Initialize the trace if necesary
-	var trace []Context
-	if config.ProofMode {
-		trace = make([]Context, 0)
-	}
+
+	trace := make([]Context, 0)
```
Comment (rodrigo-pino) [inline pkg/vm/vm.go:95]: The same here

File: cmd/cli/main.go [hunk 4]
```diff
@@ -115,24 +115,26 @@ func main() {
 						}
 					}
 
+					runner.EndRun()
```
Comment (rodrigo-pino) [inline cmd/cli/main.go:118]: I was looking into the code of `EndRun`. Outside of the scope of the PR but could you remove the `if` conditional it has inside. It is usually not a good practice to have a conditional inside a function, it's better to have it outside. For two reasons:
* Is easy for people to understand that `EndRun` only makes sense in Proof Mode
* You remove the performance cost of making a call that does nothing, when there is not proof mode. (Although, in this case, the function is probably getting inlined).

cc: @MaksymMalicki

File: cmd/cli/main.go [hunk 5]
```diff
@@ -115,24 +115,26 @@ func main() {
...
-						if traceLocation != "" {
-							if err := os.WriteFile(traceLocation, trace, 0644); err != nil {
-								return fmt.Errorf("cannot write relocated trace: %w", err)
-							}
+					}
+
+					trace, memory, err := runner.BuildProof()
+					if err != nil {
+						return fmt.Errorf("cannot build proof: %w", err)
+					}
```
Comment (rodrigo-pino) [inline cmd/cli/main.go:129]: In line with the previous comments, we shouldn't do this unless specified by the user, so it's good to have a flag here. Either proof mode is on or they asked for a trace.

File: pkg/vm/vm.go (hunk 1 above)
Comment (TAdev0) [reply pkg/vm/vm.go:526]: done!

File: cmd/cli/main.go (hunk 5 above)
Comment (TAdev0) [reply cmd/cli/main.go:132]: done

File: integration_tests/BenchMarks.txt
Comment (rodrigo-pino) [inline integration_tests/BenchMarks.txt:1]: I know this is out of scope for this PR, but you should really compare benchmarks against the Rust VM

**Review by rodrigo-pino — APPROVED** [review]

File: integration_tests/BenchMarks.txt
Comment (TAdev0) [reply integration_tests/BenchMarks.txt:1]: Well I proposed to do it and the conclusion was that it's not necessary

Comment (TAdev0) [reply integration_tests/BenchMarks.txt:1]: It was concluded that we can just benchmark locally with the rust one

**Comment by TAdev0** [comment]
I'll merge this and open a new issue for the 1 memory cell / 1 step divergence

**Pull request merged**